### DISCUSSION TITLE
Keycloak/OIDC Sync or Migration

### DISCUSSION THREAD

**Discussion by karpfediem** [kanidm/kanidm#2349] [General]
Hey, 
the software advertises itself as being a replacement for Keycloak.
As a current Keycloak user i would like to know if there is any (possibly planned) option for syncing users from Keycloak. I would like to run these systems in parallel.
In the book (https://kanidm.github.io/kanidm/stable/sync/concepts.html) i can only find options for syncing from LDAP and FreeIPA.
I couldn't find any guides on how to migrate data from Keycloak either.

I feel like you should also focus on being able to synchronize with other OIDC Identity Providers (and in particular Keycloak). 
Or if that's not feasable at least provide resources on how to perform migration from these systems.

Thanks for this project, it looks to address many common issues and specific needs I was looking to solve. Maybe I'm just missing something. Any help or pointers are appreciated :)

**Comment by Firstyear** [comment]
We haven't had any requests for it yet is why! Normally when people are using keycloak, they are using it in front of something like freeipa or ldap, meaning that we just need to sync from the true data store. 

When you say "sync from keycloak" that's a bit broader becuase keycloak does a lot of things. 

The general outline would be:

* Identify the user storage backend (sqlite, mysql, something else)
* Write a sync tool for kanidm-keycloak-<provider>-sync
** These sync tools work by reading the entries and transformining them into SCIM which then is sent to Kanidm
* Run the sync tool in parallel while you move OIDC services from keycloak to kanidm
** We can't automatically important the OIDC client definitions because they likely use different parameters than we expect and enforce so this stage will take some manual work.
* Once you're happy, turn off keycloak and finalise the migration. 

If you have some rust experience it shouldn't be "too hard" to model this on the existing ldap sync tools. Otherwise, it might be a bit more work (and time) before we get to this because we have a variety of different priorities atm.

**Reply by karpfediem** [reply]
Haha, thats fair. We just recently adopted Keycloak for our company and it did fulfill nearly all of our needs (except for Linux logins).

Thanks for the run-down though. I do have some rust experience and will check out the existing sync tools to see if i can manage :)

**Comment by anatol-newton** [comment]
Just for completeness: I just started implementing a such sync tool. As my time is pretty limited this might still take some months to be completed (or I might even abort it, if it turns out to be too much work). Ideas and help is appriciated :smile: 
https://github.com/anatol-newton/kanidm/tree/keycloak-sync

**Reply by Firstyear** [reply]
That looks great! I'll have to take a look, and if it develops further we could consider adding it to the main project too as I think this would have some broader interest.

**Reply by anatol-newton** [reply]
I sadly just learnt that exporting password hashes / credentials from Keycloak is only possible through the Keycloak CLI to file exporter. As not having to get new credentials is kind of a huge deal, I will probably resort to writing just a one-time importer from a Keycloak file export to Kanidm, unless someone has a better idea.

**Reply by Firstyear** [reply]
Hmmm that is annoying. A one time importer is a good option, you could probably still use our sync api for it and basicly one-shot the import followed by an immediate "finalise". 

Perhaps a feature request to keycloak to allow them to export these data?

**Reply by karpfediem** [reply]
When i first thought about this I considered accessing the database directly.
I can see why they would not want API based access to credential data.

**Reply by anatol-newton** [reply]
I never thought about accessing the database itself, but i guess that would be an even bigger challange than importing from a one-time file export.
And regarding the Admin REST API: I assume they do this for security reasons and won't change it. I might ask for an opt-in config, but that is probably not going to happen neither.
I'll go with the one-time migration/sync solution, as this also serves the purpose I want this feature for.

Thank you for your ideas!
